### PR TITLE
Improve classification of class/enum/struct/union

### PR DESCRIPTION
**Pull request by parkerjs**
This commit aims to improve classification of chunks pertaining to all
'non-typedef' user-defined types (classes, enums, structs, & unions).
By improving the correctness assigned to a chunk's type, parent type,
and associated flags, the accuracy and expected behavior of the code
formatting should improve, in particular for more nuanced and complex
class and structure definitions inherently allowed by the C++ language.

Some of the primary issues this commit attempts to address include the
following:

1) Improved identification of associated variable definitions,
   including declarations that may trail the actual body of the
   class/struct definition
2) Better detection of the chunk associated with actual name of the
   user-defined type
3) Better detection of attributes, declspecs, enum base specifiers,
   inheritance lists, macros, macro-like functions (which may precede
   the actual name of the user-defined type)
4) Improved marking of chunk flags, including (but not limited to):
   - PCF_LVALUE
   - PCF_VAR_TYPE
   - PCF_VAR_DEF
   - PCF_VAR_1ST
   - PCF_IN_CLASS_BASE
   - PCF_IN_CLASS
   - PCF_IN_STRUCT

As part of this commit, a new EnumStructUnionParser class has been
created and primarily supersedes the former fix_enum_struct_union()
function. Some code from other cpp files and functions has been
consolidated into this new class, while other lines of code
have been invalidated/rendered obsolete. Some re-factoring was
necessary to produce the improved chunk classification.

The improvements afforded by the changes in this commit are
particularly inherent in the following file:

- tests/cli/output/class_enum_struct_union.csv

When compared with the version of this file prior to the changes,
the improvement in classification is readily apparent. A non-csv,
traditional text version of this file has been added as part of this
commit.

As a consequence of this update, additional expected test output
files have been updated, including 'tests/cli/output/p.txt'.
For instance, the PCF_LVALUE flag was previously incorrectly
assigned to two chunks associated with a structure definition. This
commit fixes that issue, among others.

### CODE REVIEW

**Comment by parkerjs** [comment]
Compare the following files to see the improvements. The file with the 'before' suffix represents output for the the current master version of Uncrustify, while 'after' represents the output generated with my version. It may be easier to see the chunk flags by viewing the second set of csv files in the zip file.

[class_enum_struct_union_before.txt](https://github.com/uncrustify/uncrustify/files/5528026/class_enum_struct_union_before.txt)
[class_enum_struct_union_after.txt](https://github.com/uncrustify/uncrustify/files/5528027/class_enum_struct_union_after.txt)

[csv.zip](https://github.com/uncrustify/uncrustify/files/5528046/csv.zip)

File: tests/cli/test_cli_options.py [hunk 1]
```diff
@@ -726,6 +737,7 @@ def main(args):
                                   reg_replace(r', \[[_|,|1|A-Z]*\]', ', []'),
                                   reg_replace(r', \[0[xX][0-9a-fA-F]+:[_|,|1|A-Z]*\]', ', []'),
                                   reg_replace(r'   \[0[xX][0-9a-fA-F]+:[_|,|1|A-Z]*\]', '   []'),
+                                  reg_replace(r'^[ \t]*[_A-Za-z][_A-Za-z0-9]*::', ''),
```
Comment (parkerjs) [inline tests/cli/test_cli_options.py:740]: This should've resolved the CI build failure in 'tests/cli/output/92.txt'...

Comment (gmaurel) [reply tests/cli/test_cli_options.py:740]: fine work. push it up

**Comment by gmaurel** [comment]
I had this problem also.
The solution is written at file log_rules.cpp line 36...
 ```.cpp
  // some Windows provide "ABC::XYZ::function_Name" as __func__
   // we look for the last ':' character
   // a function rindex cannot be found
```
You can copy this part of the code

**Comment by gmaurel** [comment]
Have a look to the log:
...
```.log
- EnumStructUnionParser::try_find_end_chunk : orig_line is 1, orig_col is 8, pc->text() 'TelegramIndex'
+ try_find_end_chunk : orig_line is 1, orig_col is 8, pc->text() 'TelegramIndex'
```

File: src/logger.cpp [hunk 2]
```diff
@@ -312,3 +312,24 @@ void log_func_stack(log_sev_t sev, const char *prefix, const char *suffix, size_
...
+   // a function rindex cannot be found
+   size_t length_of_string = strlen(func);
+
+   for (int i = length_of_string - 1; i > 0; --i)
+   {
+      if (func[i] == ':')
+      {
+         return(func + i);
+      }
+   }
```
Comment (parkerjs) [inline src/logger.cpp:331]: Oops, not all control paths return a value. I'll fix and re-upload.

**Comment by gmaurel** [comment]
@parkerjs 
There is still a problem:
Problem (2) with results\92.txt
Please have a look to:
  https://ci.appveyor.com/project/gmaurel/uncrustify/builds/36360817/job/mif3hxn9a948d72v

at line 2936, ... 3102:
```.log
"- EnumStructUnionParser::try_find_end_chunk : orig_line is 1, orig_col is 8, pc->text() 'TelegramIndex'"
'? -----------------------\n'
"+ try_find_end_chunk : orig_line is 1, orig_col is 8, pc->text() 'TelegramIndex'"
'     pc->type is TYPE, pc->parent_type is NONE => *type is WORD, *parent_type is NONE'
"  make_type : orig_line is 1, orig_col is 8, pc->text() 'TelegramIndex'"
'     pc->type is WORD, pc->parent_type is NONE => *type is TYPE, *parent_type is NONE'
"- EnumStructUnionParser::mark_constructors : orig_line is 3, orig_col is 1, pc->text() 'TelegramIndex'"
'? -----------------------\n'
"+ mark_constructors : orig_line is 3, orig_col is 1, pc->text() 'TelegramIndex'"
```

File: src/logger.h [hunk 3]
```diff
@@ -124,7 +124,7 @@ void log_flush(bool force_nl);
  * It uses the log_func class to add an entry to the function log stack.
  * It is automatically removed when the function returns.
  */
-#define LOG_FUNC_ENTRY()    log_func log_fe = log_func(__func__, __LINE__)
+#define LOG_FUNC_ENTRY()    log_func log_fe = log_func(get_unqualified_func_name(__func__), __LINE__)
```
Comment (parkerjs) [inline src/logger.h:127]: I hope this fixes the problem in results\92.txt

**Comment by gmaurel** [comment]
we have still a warning. Please have a look to:
  https://ci.appveyor.com/project/uncrustify/uncrustify/builds/36392480/job/e76y5s96u5c23csk?fullLog=true

at line 81:
```
c:\projects\uncrustify\src\EnumStructUnionParser.h(466): warning C4099: 'std::map<c_token_t,std::map<size_t,chunk_t *,std::less<_Kty>,std::allocator<std::pair<const _Kty,_Ty>>>,std::less<c_token_t>,std::allocator<std::pair<const c_token_t,std::map<_Kty,_Ty,std::less<_Kty>,std::allocator<std::pair<const _Kty,_Ty>>>>>>': type name first seen using 'class' now seen using 'struct' [C:\projects\uncrustify\build\uncrustify.vcxproj]
          with
          [
              _Kty=size_t,
              _Ty=chunk_t *
          ]
  c:\projects\uncrustify\src\EnumStructUnionParser.h(466): note: see declaration of 'std::map<c_token_t,std::map<size_t,chunk_t *,std::less<_Kty>,std::allocator<std::pair<const _Kty,_Ty>>>,std::less<c_token_t>,std::allocator<std::pair<const c_token_t,std::map<_Kty,_Ty,std::less<_Kty>,std::allocator<std::pair<const _Kty,_Ty>>>>>>'
          with
          [
              _Kty=size_t,
              _Ty=chunk_t *
          ]
```

**Review by gmaurel — CHANGES_REQUESTED** [review]
see my comment about the warning

**Comment by gmaurel** [comment]
look at:
https://alfps.wordpress.com/2010/06/22/cppx-is-c4099-really-a-sillywarning-disabling-msvc-sillywarnings/

**Comment by gmaurel** [comment]
https://stackoverflow.com/questions/468486/warning-c4099-type-name-first-seen-using-class-now-seen-using-struct-ms-vs
...

1

In c++ the only difference between a class and a struct is that class's member variables, member functions and base classes are private by default, while in a struct they're by default public; so, the fact that the class is POD should not make any difference here.

**Comment by gmaurel** [comment]
Using 
```.log
The C compiler identification is MSVC 19.27.29111.0
The CXX compiler identification is MSVC 19.27.29111.0
```
The wrning is no more present.

**Review by gmaurel — APPROVED** [review]

**Pull request merged**